api/src/main/scala/dx/api/DxApi.scala

```diff
@@ -938,4 +942,57 @@ case class DxApi(version: String = "1.0.0", dxEnv: DXEnvironment = DXEnvironment
     val path = FileUtils.writeFileContent(tempFile, content)
     uploadFile(path, Some(destination), wait = wait)
   }
+
+  private case class UploadFileVisitor(sourceDir: Path,
+                                       destination: Option[String],
+                                       waitOnUpload: Boolean)
+      extends SimpleFileVisitor[Path] {
+    private def ensureEndsWithSlash(path: String): String = {
+      if (path.endsWith("/")) path else s"${path}/"
+    }
+
+    val (projectId: Option[String], folder: String) = destination
+      .map { dest =>
+        dest.split(":").toVector match {
+          case Vector(projectId) if projectId.startsWith("project-") =>
+            (Some(projectId), "/")
+          case Vector(folder) =>
+            (None, ensureEndsWithSlash(folder))
+          case Vector(projectId, folder) =>
+            (Some(projectId), ensureEndsWithSlash(folder))
+          case _ =>
+            throw new Exception(s"invalid destination ${dest}")
+        }
+      }
+      .getOrElse((None, "/"))
+
+    private var uploadedFiles = Vector.empty[DxFile]
+
+    def result: (Option[String], String, Vector[DxFile]) = {
+      (projectId, folder, uploadedFiles)
+    }
+
+    override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
+      val fileRelPath = sourceDir.relativize(file)
+      val fileDestPath = s"${folder}${fileRelPath}"
+      val fileDest = projectId.map(p => s"${p}:${fileDestPath}").getOrElse(fileDestPath)
+      val dxFile = uploadFile(file, Some(fileDest), waitOnUpload)
+      uploadedFiles :+= dxFile
+      FileVisitResult.CONTINUE
+    }
+  }
+
+  /**
+    * Uploads all the files in a directory, and - if `recursive=true` - all
+    * files in subfolders as well. Returns (projectId, folder, files).
+    */
+  def uploadDirectory(path: Path,
```

Comment:
Is there a reason not to just use `dx upload -r`, like you don't get back the information you need about each file?

Reply:
Good question. In `dx upload` recursive is done by just looping over all the files serially (i.e. there is no parallelization, see https://github.com/dnanexus/dx-toolkit/blob/master/src/python/dxpy/scripts/dx.py#L2050). Since `uploadFile` is already working I thought it would be better to leverage that and be able to return the information for all the uploaded files. You could implement it with `dx upload -r` but then you either couldn't return the file information, you'd have to do a separate query, or you'd have to parse the output of `dx upload -r`.